src/transformStateWithClones.js

```diff
@@ -7,7 +7,29 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+    let currentState = { ...state };
+    const stateHistory = [];
+
+    actions.forEach(action => {
+        switch (action.type) {
+            case 'clear':
+                currentState = {};
+                break;
+            case 'addProperties':
+                currentState = { ...currentState, ...action.extraData };
+                break;
+            case 'removeProperties':
+                action.keysToRemove.forEach(key => {
+                    delete currentState[key];
+                });
+                break;
+            default:
+                break;
```

Comment:
The use of a `switch` statement is good practice here as it adheres to the [CODE STYLE] checklist item. However, ensure that the `default` case provides meaningful error handling or logs an error message, as it's important for debugging and understanding what went wrong when an action with an unexpected type is encountered.